src/solvers/refinement/string_refinement_util.h

```diff
@@ -0,0 +1,315 @@
+class equation_symbol_mappingt
+{
+public:
+  /// Record the fact that equation `i` contains expression `expr`
+  void add(const std::size_t i, const exprt &expr);
+
+  /// \param i: index corresponding to an equation
+  /// \return vector of expressions contained in the equation with the given
+  ///   index `i`
+  std::vector<exprt> find_expressions(const std::size_t i);
+
+  /// \param expr: an expression
+  /// \return vector of equations containing the given expression `expr`
+  std::vector<std::size_t> find_equations(const exprt &expr);
+
+private:
+  /// Record index of the equations that contain a given expression
+  std::map<exprt, std::vector<std::size_t>> equations_containing;
+  /// Record expressions that are contained in the equation with the given index
+  std::unordered_map<std::size_t, std::vector<exprt>> strings_in_equation;
+};
+
+/// Base class for string primitives
+class string_primitivet
+{
+public:
+  string_primitivet() = default;
+  string_primitivet(const string_primitivet &) = delete;
+
+  virtual optionalt<array_string_exprt> string_result() const
+  {
+    return {};
+  }
+
+  virtual std::vector<array_string_exprt> string_arguments() const
+  {
+    return {};
+  }
+};
+
+/// String primitive transforming one string into another
+class string_transformation_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  array_string_exprt input;
+  std::vector<exprt> args;
+  exprt return_code;
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return {input};
+  }
+};
+
+/// String inserting a string into another one
+class string_insertion_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  array_string_exprt input1;
+  array_string_exprt input2;
+  std::vector<exprt> args;
+  exprt return_code;
+
+  /// Constructor from arguments of a function application
+  string_insertion_primitivet(
+    const std::vector<exprt> &fun_args,
+    array_poolt &array_pool);
+
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return {input1, input2};
+  }
+};
+
+/// String creation from other types
+class string_creation_primitivet : public string_primitivet
+{
+public:
+  array_string_exprt result;
+  std::vector<exprt> args;
+  exprt return_code;
+
+  optionalt<array_string_exprt> string_result() const override
+  {
+    return result;
+  }
+};
+
+/// String test
+class string_test_primitivet : public string_primitivet
+{
+public:
+  exprt result;
+  std::vector<array_string_exprt> under_test;
+  std::vector<exprt> args;
+  std::vector<array_string_exprt> string_arguments() const override
+  {
+    return under_test;
+  }
+};
+
+/// Keep track of dependencies between strings.
+/// Each string points to primitive calls on which it depends,
```

Comment:
What is a primitive call? You mean one of the intrinsics provided by the string solver, like __CPROVER_string_equal et al?

Reply:
yes, but I'm open to suggestion for a better vocabulary.

Reply:
I'd probably call them "string intrinsics", since "primitives" makes me think of primitive types

Reply:
@smowton What about "builtin functions" ? I would rename `string_primitivet` to `string_builtin_functiont` and so on.

Reply:
They're not quite builtins, they belong to the string solver. So `string_solver_function_callt`?